### PR TITLE
Dynamic: Share more regexp code 

### PR DESCRIPTION
**Pull request by erik-krogh**
In preparation for a query that can flag CVE-2022-31163

I've some query code for regex based queries into the shared `regex` pack.   

The hostname code was shared using the `sync-files.py` script.  

The regex/missing-regexp-anchor queries were shared using copy-pasted code that were almost identical, but not quite.  
So we could have risked that the implementation came out of sync in the future.  
  
The signature `MissingRegExpAnchorSig` ended up being a bit complicated (two module parameters in the signature).  
But that is the clean solution, where no signature has more features than they need.  
An alternative would be to combine `HostnameRegexpSig` and `MissingRegExpAnchorSig` into one, but I that doesn't seem like a "clean" solution to me.  

The `getACommonTld()` predicate was copy-pasted between the different implementations, but there was no check that the implementations were the same, so I moved it into the shared pack.  

The change-note is already covered by the "the regex implementations got moved to a shared pack"-change-note.

### CODE REVIEW

File: python/ql/lib/semmle/python/RegexTreeView.qll [hunk 1]
```diff
@@ -829,6 +848,19 @@ module Impl implements RegexTreeViewSig {
...
+   *
+   * Example:
+   *
+   * ```
+   * \A
+   * ```
+   */
+  class RegExpAnchor extends RegExpSpecialChar {
+    RegExpAnchor() { this.getChar() = ["^", "$"] }
+  }
```
Comment (yoff) [inline python/ql/lib/semmle/python/RegexTreeView.qll:862]:
```suggestion
  /**
   * A term that matches a specific position between characters in the string.
   *
   * Example:
   *
   * ```
   * \A
   * ```
   */
  class RegExpAnchor extends RegExpSpecialChar {
    RegExpAnchor() { this.getChar() = ["\\A", "^", "$", "\\Z"] }
  }
```

File: shared/regex/codeql/regex/HostnameRegexp.qll [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    or
+    exists(RegExpTerm parent |
+      term = getLastChild(parent) and
+      isRightArmTerm(parent)
+    )
+  }
+
+  private RegExpTerm getLastChild(RegExpTerm parent) {
+    result = max(RegExpTerm child, int i | child = parent.getChild(i) | child order by i)
+  }
```
Comment (yoff) [inline shared/regex/codeql/regex/HostnameRegexp.qll:255]: Do you need to implement this? It looks like it is added to the signature.

File: shared/regex/codeql/regex/HostnameRegexp.qll [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+  /** Holds if `term` is one of the transitive right children of a regexp. */
+  predicate isRightArmTerm(RegExpTerm term) {
+    term.isRootTerm()
+    or
+    exists(RegExpTerm parent |
+      term = getLastChild(parent) and
+      isRightArmTerm(parent)
+    )
+  }
```
Comment (yoff) [inline shared/regex/codeql/regex/HostnameRegexp.qll:251]: I am not sure why these are in this module, they are not needed here and not specific to host names. I would expect them either in the specific module where they are needed (MissingRegexpAnchor) or in a general one, if we think they are generally useful. Do we expect them to be used a lot for host names?

**Review by yoff — CHANGES_REQUESTED** [review]
Looks great, only minor comments. It appears that Python could gain the `MissingRegExpAnchor` query quite easily now, perhaps even as part of this PR.

File: shared/regex/codeql/regex/HostnameRegexp.qll (hunk 3 above)
Comment (erik-krogh) [reply shared/regex/codeql/regex/HostnameRegexp.qll]: The predicates are used in multiple locations, and I needed them to be somewhere.  

They feel too specific to add to the `RegexTreeView` signature (that just would just further increase the amount of code each language needs to implement).   
Yes, I could add something like `RegexUtils.qll` file to the shared pack, but isn't that a bit overkill?  
(I'll do it if you think it's fine).

Comment (yoff) [reply shared/regex/codeql/regex/HostnameRegexp.qll]: I was thinking about a `RegexUtil.qll` but if we do not have one already, then I think it is fine either way.

**Review by yoff — APPROVED** [review]
LGTM

File: shared/regex/codeql/regex/HostnameRegexp.qll (hunk 3 above)
Comment (erik-krogh) [reply shared/regex/codeql/regex/HostnameRegexp.qll]: We do have `NfaUtils.qll`, which is close, but these predicates don't have anything to do with NFAs, so I don't think they belong there.

Comment (yoff) [reply shared/regex/codeql/regex/HostnameRegexp.qll]: Ah, that was the one. Yeah, let us keep them here for now, and create `RegexUtils` once we have a bit more utility code..

**Pull request merged**